src/process-events.c

```diff
@@ -263,22 +258,136 @@ static __always_inline void push_telemetry_event(struct pt_regs *ctx, ptelemetry
     else                                                                    \
     {                                                                       \
-        if (count > 0)                                                      \
-        {                                                                   \
-            ev->u.v.truncated = FALSE;                                      \
-            push_telemetry_event(ctx, ev);                                  \
-        }                                                                   \
+        /* save it here since push_telemetry_event clears ev */             \
+        br = ev->u.v.value[0];                                              \
+        ev->u.v.truncated = FALSE;                                          \
+        push_telemetry_event(ctx, ev);                                      \
+        if (br == '/')                                                      \
+            goto Skip;                                                      \
         /* we're done here, follow the pointer */                           \
+        void *old_ptr = ptr;                                                \
         bpf_probe_read(&ptr, sizeof(ptr), ptr + parent);                    \
-        to_skip += 1;                                                       \
-        if (!ptr)                                                           \
-            goto Skip;                                                      \
-        if (br == '/')                                                      \
+        if (!ptr || old_ptr == ptr)                                         \
             goto Skip;                                                      \
+        to_skip += 1;                                                       \
         br = 0;                                                             \
     }
 
-#define SEND_PATH_N(N) REPEAT_##N(SEND_PATH;)
+#define SEND_PATH_N(N, T) REPEAT_##N(SEND_PATH(T);)
+
+static __always_inline bool check_discard(struct pt_regs *ctx, ptelemetry_event_t ev, u64 *id)
+{
+    u32 retcode = (u32)PT_REGS_RC(ctx);
+    if (retcode != 0) {
+        ev->id = *id;
+        ev->telemetry_type = TE_DISCARD;
+        ev->u.r.retcode = retcode;
+        push_telemetry_event(ctx, ev);
+
+        return true;
+    }
+
+    return false;
+}
+
+static __always_inline void* get_current_exe()
+{
+    u64 offset_key = CRC_LOADED;
+    void *offset = bpf_map_lookup_elem(&offsets, &offset_key);
+    if (!offset)
+        return NULL;
+
+    void *ts = (void *)bpf_get_current_task();
+    if (!ts)
+        return NULL;
+
+    void *ptr = NULL;
+    read_value(ts, CRC_TASK_STRUCT_MM, &ptr, sizeof(ptr));
+    read_value(ptr, CRC_MM_STRUCT_EXE_FILE, &ptr, sizeof(ptr));
+
+    return ptr;
+}
+
+// it's argument should be a pointer to a file
+static __always_inline file_info_t extract_file_info(void *ptr)
```

Comment:
Ditto

Reply:
As of right now no one else is calling for `bpf_get_current_task` so I rather not extract it outside this file until it is actually common across different telemetry sources.